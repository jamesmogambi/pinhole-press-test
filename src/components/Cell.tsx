import React from "react";
import cellIcon from "../icon.svg";

type CellProp = {
  value: string | number;
  icon: boolean;
  styleProp: any;
};

const Cell: React.FC<CellProp> = ({ value, styleProp, icon }) => {
  return (
    <div className={`${styleProp} cell`}>
      {icon && <img src={cellIcon} alt="icon" className="icon" />}
      {value}
    </div>
  );
};

export default Cell;
