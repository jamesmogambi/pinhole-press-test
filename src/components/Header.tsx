import React from "react";

const Header: React.FC<{ states: IState[] }> = ({ states }) => {
  return (
    <React.Fragment>
      <div className="header">
        <div className="container shadow">
          <span className="title">PRODUCTS</span>
          <div className="items">
            <span className="item">Category</span>
            <span className="item">Sub-Category</span>
          </div>
        </div>
        <div className="container">
          <span className="title title-state">STATE</span>
          <div className="items">
            {states.map((s, i) => (
              <span key={i} className="item">
                {s.state}
              </span>
            ))}
          </div>
        </div>
      </div>
    </React.Fragment>
  );
};

export default Header;
