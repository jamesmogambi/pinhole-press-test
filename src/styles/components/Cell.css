.cell {
  width: 120px;
  font-size: 15px;
  padding: 10px;
  color: rgb(78, 78, 78);
  margin: 0px;
}

.cell-dark {
  background-color: var(--dark-gray-color);
}

.cell-light {
  background-color: var(--white-color);
  text-align: center;
}

.text-dark {
  color: #000000;
  font-weight: bold;
}

.text-light {
  color: var(--dark-gray-color);
}

.cell-double {
  width: 240px;
}

.cell-total {
  text-align: center;
}

.icon {
  height: 20px;
  margin-right: 10px;
}

.cell-icon {
  display: flex;
}
