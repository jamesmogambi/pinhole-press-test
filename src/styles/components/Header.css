.header {
  background: var(--dark-blue-color);
  height: 105px;
  display: flex;
  padding-top: 7px;
}

.title {
  color: var(--white-color);
  font-weight: bold;
  padding-left: 10px;
}

.item {
  color: rgba(250, 235, 215, 0.774);
  width: 120px;
  font-size: 15px;
  padding: 10px;
}

.items {
  margin-top: 20px;
  display: flex;
  flex-direction: row;
}

.title-state {
  position: relative;
  left: 8%;
}
