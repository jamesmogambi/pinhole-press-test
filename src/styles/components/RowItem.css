.row-items,
.column {
  display: flex;
}

.row-items {
  flex-direction: row;
}

.column {
  flex-direction: column;
}
