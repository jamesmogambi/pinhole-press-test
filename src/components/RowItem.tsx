import React from "react";
import { connect } from "react-redux";
import {
  getSubCategories,
  getSubTotal,
  getTotal,
} from "../selectors/categories";
import Cell from "./Cell";

interface RowItemProp {
  category: Category;
  states: IState[];
  sales: ISale[];
  subcategories: SubCategory[];
}

export const RowItem: React.FC<RowItemProp> = (props) => {
  let { category, subcategories, states, sales } = props;

  const renderSubTotal = (item: SubCategory) => {
    return states.map((s, i) => {
      let { state } = s;
      let subTotal = getSubTotal(item, state, sales);
      return (
        <Cell styleProp="cell-light" key={i} value={subTotal} icon={false} />
      );
    });
  };

  const renderTotal = () => {
    return states.map((s, i) => {
      let { state } = s;
      let total = getTotal(state, category, sales);
      return (
        <Cell
          key={i}
          styleProp="text-dark cell-dark cell-total"
          value={total}
          icon={false}
        />
      );
    });
  };
  return (
    <div>
      <div className="row-items">
        <Cell
          value={category.name}
          styleProp="text-dark cell-icon"
          icon={true}
        />
        <div className="column">
          {subcategories.map((item, i) => (
            <div key={i} className="row-items">
              <Cell styleProp={null} value={item.subCategory} icon={false} />
              <div className="row-items">{renderSubTotal(item)}</div>
            </div>
          ))}
        </div>
      </div>
      <div className="row-items">
        <Cell
          styleProp="text-dark cell-double"
          value={`${category.name} total`}
          icon={false}
        />
        <div className="row-items">{renderTotal()}</div>
      </div>
    </div>
  );
};

const mapStateToProps = (state: any, ownProps: any) => {
  let { sales } = state.sale;
  let { category } = ownProps;
  return {
    sales,
    subcategories: getSubCategories(category, sales),
  };
};

export default connect(mapStateToProps)(RowItem);
