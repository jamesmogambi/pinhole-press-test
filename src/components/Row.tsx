import React from "react";
import RowItem from "./RowItem";
import Cell from "./Cell";
import { getGrandTotal } from "../selectors/categories";

interface RowProp {
  states: IState[];
  sales: ISale[];
  categories: Category[];
}

export const Row: React.FC<RowProp> = ({ states, categories, sales }) => {
  const renderGrandTotal = () => {
    return states.map((s, i) => {
      let { state } = s;
      let grandTotal = getGrandTotal(state, sales);
      return (
        <Cell
          key={i}
          styleProp="text-dark cell-dark cell-total"
          value={grandTotal}
          icon={false}
        />
      );
    });
  };

  return (
    <React.Fragment>
      <div className="container">
        {categories.map((item, i) => (
          <RowItem key={i} category={item} states={states} />
        ))}
        <div className="row-items item-border">
          <Cell
            styleProp="text-dark cell-double"
            value="Grand Total"
            icon={false}
          />
          {renderGrandTotal()}
        </div>
      </div>
    </React.Fragment>
  );
};

export default Row;
