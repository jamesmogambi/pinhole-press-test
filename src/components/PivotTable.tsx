import React from "react";
import { connect } from "react-redux";
import selectStates from "../selectors/states";
import { getCategories } from "../selectors/categories";
import Header from "./Header";
import Row from "./Row";

interface PivotTableProps {
  sales: ISale[];
  states: IState[];
  categories: Category[];
}
export const PivotTable: React.FC<PivotTableProps> = (props) => {
  return (
    <React.Fragment>
      <div className="scroll bg">
        <Header states={props.states} />
        <Row
          states={props.states}
          categories={props.categories}
          sales={props.sales}
        />
      </div>
    </React.Fragment>
  );
};

const mapStateToProps = (state: any) => {
  let { sales } = state.sale;
  return {
    states: selectStates(sales),
    sales,
    categories: getCategories(sales),
  };
};

export default connect(mapStateToProps)(PivotTable);
