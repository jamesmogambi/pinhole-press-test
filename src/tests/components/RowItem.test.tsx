import { shallow } from "enzyme";
import { RowItem } from "../../components/RowItem";
import sales from "../fixtures/sales";

test("should correctly render RowItem with category ,subcategories , subtotal and total", () => {
  const subcategories = [
    { category: "Furniture", subCategory: "Bookcases" },
    { category: "Furniture", subCategory: "Chairs" },
  ];

  const states = [
    { state: "Alabama", city: "Decatur" },
    { state: "Arizona", city: "Gilbert" },
    { state: "Arkansas", city: "Fayetteville" },
  ];

  let wrapper = shallow(
    <RowItem
      category={{ name: "Furniture" }}
      states={states}
      subcategories={subcategories}
      sales={sales}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
