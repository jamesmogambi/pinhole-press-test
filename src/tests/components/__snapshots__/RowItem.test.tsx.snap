// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render RowItem with category ,subcategories , subtotal and total 1`] = `
<div>
  <div
    className="row-items"
  >
    <Cell
      icon={true}
      styleProp="text-dark cell-icon"
      value="Furniture"
    />
    <div
      className="column"
    >
      <div
        className="row-items"
        key="0"
      >
        <Cell
          icon={false}
          styleProp={null}
          value="Bookcases"
        />
        <div
          className="row-items"
        >
          <Cell
            icon={false}
            key="0"
            styleProp="cell-light"
            value={1694}
          />
          <Cell
            icon={false}
            key="1"
            styleProp="cell-light"
            value={0}
          />
          <Cell
            icon={false}
            key="2"
            styleProp="cell-light"
            value={0}
          />
        </div>
      </div>
      <div
        className="row-items"
        key="1"
      >
        <Cell
          icon={false}
          styleProp={null}
          value="Chairs"
        />
        <div
          className="row-items"
        >
          <Cell
            icon={false}
            key="0"
            styleProp="cell-light"
            value={0}
          />
          <Cell
            icon={false}
            key="1"
            styleProp="cell-light"
            value={15}
          />
          <Cell
            icon={false}
            key="2"
            styleProp="cell-light"
            value={0}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="row-items"
  >
    <Cell
      icon={false}
      styleProp="text-dark cell-double"
      value="Furniture total"
    />
    <div
      className="row-items"
    >
      <Cell
        icon={false}
        key="0"
        styleProp="text-dark cell-dark cell-total"
        value={1694}
      />
      <Cell
        icon={false}
        key="1"
        styleProp="text-dark cell-dark cell-total"
        value={15}
      />
      <Cell
        icon={false}
        key="2"
        styleProp="text-dark cell-dark cell-total"
        value={0}
      />
    </div>
  </div>
</div>
`;
